RFS/src/main/java/org/opensearch/migrations/bulkload/common/RfsDocument.java

```diff
@@ -16,28 +23,56 @@
  */
 @AllArgsConstructor
 public class RfsDocument {
+    protected static final ObjectMapper OBJECT_MAPPER = ObjectMapperFactory.createDefaultMapper();
     // Originally set to the lucene index doc number, this number helps keeps track of progress over a work item
     public final int progressCheckpointNum;
 
     // The Elasticsearch/OpenSearch document to be reindexed
-    public final BulkDocSection document;
+    public final BulkOperationSpec document;
 
     public static RfsDocument fromLuceneDocument(RfsLuceneDocument doc, String indexName) {
-        return new RfsDocument(
-            doc.luceneDocNumber,
-            new BulkDocSection(
-                doc.id,
-                indexName,
-                doc.type,
-                doc.source,
-                doc.routing
-            )
-        );
+        // Check if this is a delete operation
+        Map<String, Object> document = null;
+        if (doc.source != null) {
+            try {
+                document = OBJECT_MAPPER.readValue(doc.source, new TypeReference<>() {});
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Failed to parse source doc: " + e.getMessage(), e);
+            }
+        }
+
+        if (doc.operation == RfsDocumentOperation.DELETE) {
```

Comment:
Kind of lame to write all that shared serialization code but we still need to resort to `if (...) { new() } else if { new() }` pattern